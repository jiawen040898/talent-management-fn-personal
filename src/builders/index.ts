export * from './base-query.builder';
export * from './employee-task-query.builder';
export * from './feedback-cycle-reviewee-query.builder';
export * from './feedback-cycle-reviewer-query.builder';
export * from './learning-goal-query.builder';
