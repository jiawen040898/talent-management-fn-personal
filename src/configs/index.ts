export * from './aws.config';
export * from './gcp.config';
export * from './recommendation.config';
