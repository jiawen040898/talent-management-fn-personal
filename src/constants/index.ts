export * from './action-source.enum';
export * from './big-queries.enum';
export * from './cache-store.enum';
export * from './company-module.enum';
export * from './company-status.enum';
export * from './due-type.enum';
export * from './employee-score.enum';
export * from './employee-task.enum';
export * from './event-type.enum';
export * from './feedback-recommendation.enum';
export * from './fit-score-recipe.enum';
export * from './person-score.enum';
export * from './psychology-api-errors.enum';
export * from './questionnaire-framework-type.enum';
export * from './sqs-event-type.enum';
export * from './unleash-errors.enum';
