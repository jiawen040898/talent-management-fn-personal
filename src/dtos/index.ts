export * from './company.dto';
export * from './cron-schedule-message.dto';
export * from './employee.dto';
export * from './employee-account-created-event.dto';
export * from './employee-assessment-score-event.dto';
export * from './employee-fit-score-event.dto';
export * from './feedback-cycle.dto';
export * from './fit-score-recipe.dto';
export * from './manager-subordinate-aggregated-report-data.dto';
