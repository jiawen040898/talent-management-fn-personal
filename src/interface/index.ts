export * from '../dtos/fit-score-recipe.dto';
export * from './cognitive-score.interface';
export * from './employee-score.interface';
export * from './event-model.interface';
export * from './feedback-cycle.interface';
export * from './feedback-recommendation.interface';
export * from './fit-score.interface';
export * from './personality-score.interface';
export * from './program.interface';
export * from './redis.interface';
