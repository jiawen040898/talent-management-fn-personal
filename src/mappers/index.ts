export * from './employee-score.mapper';
export * from './feedback.mapper';
export * from './feedback-recommendation.mapper';
