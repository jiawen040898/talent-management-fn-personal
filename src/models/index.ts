export * from './employee.entity';
export * from './employee-questionnaire.entity';
export * from './employee-score.entity';
export * from './employee-task.entity';
export * from './entities';
export * from './feedback-cycle.entity';
export * from './feedback-cycle-email-template.entity';
export * from './feedback-cycle-notification.entity';
export * from './feedback-cycle-reviewee.entity';
export * from './feedback-cycle-reviewer.entity';
export * from './learning-goal.entity';
export * from './participant.entity';
export * from './program.entity';
