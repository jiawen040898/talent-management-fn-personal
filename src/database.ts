import { secretService } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import { ENTITIES } from './models';

type DbSecret = {
    username: string;
    engine: string;
    port: number;
    host: string;
    password: string;
    schema: string;
    dbname: string;
    dbInstanceIdentifier: string;
};

let dataSource: DataSource;

export const getDataSource = async (): Promise<DataSource> => {
    if (dataSource?.isInitialized) {
        return dataSource;
    }

    const dbSecret = await secretService.getSecret<DbSecret>(
        process.env.SM_NAME as string,
    );

    dataSource = await new DataSource({
        type: 'postgres',
        host: dbSecret.host,
        port: dbSecret.port,
        username: dbSecret.username,
        password: dbSecret.password,
        database: dbSecret.dbname,
        schema: dbSecret.schema,
        entities: [...ENTITIES],
        maxQueryExecutionTime: 1000,
        synchronize: false,
        migrationsRun: false,
        migrations: [],
        extra: {
            connectionTimeoutMillis: 3000,
            idleTimeoutMillis: 1000,
            max: 1,
        },
        logging: ['error'],
    }).initialize();

    return dataSource;
};
