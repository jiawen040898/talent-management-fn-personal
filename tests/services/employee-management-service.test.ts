import { SNSClient } from '@aws-sdk/client-sns';
import { DataSource, Repository } from 'typeorm';

import * as database from '../../src/database';
import { Employee, EmployeeTask, LearningGoal } from '../../src/models';
import { EmployeeManagementService } from '../../src/services/employee-management.service';
import { getTestDataSource } from '../setup';
import { testData } from './fixtures/test-data';

describe('EmployeeManagementService', () => {
    let dataSource: DataSource;
    let employeeRepo: Repository<Employee>;
    let learningGoalRepo: Repository<LearningGoal>;
    let employeeTaskRepo: Repository<EmployeeTask>;

    const spySNSClient = jest
        .spyOn(SNSClient.prototype, 'send')
        .mockImplementation(jest.fn());

    beforeAll(async () => {
        dataSource = await getTestDataSource();

        employeeRepo = dataSource.getRepository(Employee);
        learningGoalRepo = dataSource.getRepository(LearningGoal);
        employeeTaskRepo = dataSource.getRepository(EmployeeTask);

        await employeeRepo.save(testData.employeeList);
        await learningGoalRepo.save(testData.learningGoals);
        await employeeTaskRepo.save(testData.employeeTasks);

        jest.spyOn(database, 'getDataSource').mockImplementation(() => {
            return Promise.resolve(dataSource);
        });
    });

    describe('updateUserAccountId', () => {
        it('should update user account id', async () => {
            // Act
            await EmployeeManagementService.updateUserAccountId(
                testData.testEmployeeUserCreatedEventData,
            );

            // Assert
            const employee = await employeeRepo.findOneBy({
                id: testData.employee1.id,
            });
            expect(employee?.user_account_id).toEqual(
                testData.testEmployeeUserCreatedEventData.id,
            );
        });

        it('should not replace employee that already have user account id', async () => {
            // Act
            await EmployeeManagementService.updateUserAccountId(
                testData.testEmployeeUserCreatedEventData2,
            );

            // Assert
            const employee = await employeeRepo.findOneBy({
                id: testData.employee2.id,
            });
            expect(employee?.user_account_id).toEqual(13);
        });
    });

    describe('getGoalAndTaskDueListByCompany', () => {
        it('should return all employees task and goals due list', async () => {
            // Act
            const actual =
                await EmployeeManagementService.getGoalAndTaskDueListByCompany(
                    testData.mockCompanyLookupData,
                );

            // Assert
            expect(actual).toMatchSnapshot();
        });
    });

    describe('processEmployeeTaskAndGoalByCompany', () => {
        it('should publish message to notification', async () => {
            // Act
            await EmployeeManagementService.processEmployeeTaskAndGoalByCompany(
                testData.mockCompanyLookupData,
            );

            // Assert
            expect(spySNSClient).toHaveBeenCalledTimes(2);
        });
    });
});
