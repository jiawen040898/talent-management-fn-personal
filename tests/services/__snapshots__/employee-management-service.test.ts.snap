// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeManagementService getGoalAndTaskDueListByCompany should return all employees task and goals due list 1`] = `
[
  EmployeeTaskAndGoalDue {
    "company_id": 5,
    "due_date": "2022-11-26 00:00:00.000",
    "due_type": "in 3 days",
    "employee_email": "peterl@gmail.com",
    "employee_first_name": "Peter",
    "employee_goals_due_count": 1,
    "employee_id": "7d7c0426-01a6-4577-ab63-83bf2e3c7655",
    "employee_tasks_due_count": 1,
    "employee_user_account_id": 13,
  },
  EmployeeTaskAndGoalDue {
    "company_id": 5,
    "due_date": "2022-11-23 00:00:00.000",
    "due_type": "today",
    "employee_email": "jasonl@gmail.com",
    "employee_first_name": "Jason",
    "employee_goals_due_count": 1,
    "employee_id": "ed071ebf-6f7d-47d1-9b0b-52005a5cf778",
    "employee_tasks_due_count": 1,
    "employee_user_account_id": 25,
  },
]
`;
